fastchat/serve/model_worker.py

```diff
@@ -184,6 +184,37 @@ def generate_stream_gate(self, params):
             }
             yield json.dumps(ret).encode() + b"\0"
 
+    def generate_completion(self, params):
+        try:
+            outputs = ""
+            finish_reason = "stop"
+            for output in self.generate_stream_func(
+                self.model,
+                self.tokenizer,
+                params,
+                self.device,
+                self.context_len,
+                1,
+            ):
+                outputs = output
+                if len(self.tokenizer(outputs).input_ids) >= params["max_tokens"]:
+                    finish_reason = "length"
+                    break
+            return json.dumps({"text": outputs, "finish_reason": finish_reason, "completion_tokens": len(self.tokenizer(outputs).input_ids), "prompt_tokens": len(self.tokenizer(params["prompt"]).input_ids)})
+        except torch.cuda.OutOfMemoryError:
+            ret = {
+                "text": server_error_msg,
+                "error_code": 1,
+            }
+            return json.dumps(ret).encode() + b"\0"
+
+    def generate_embeddings(self, params):
+        tokenizer = self.tokenizer
+        input_ids = tokenizer.encode(params["input"], return_tensors='pt').to(self.device)
+        model_output = self.model(input_ids, output_hidden_states=True)
+        data = model_output.hidden_states[-1][0]
```

Comment:
does this work for all models?

Reply:
It works for huggingface models, see [hidden_states](https://huggingface.co/docs/transformers/main_classes/output#transformers.modeling_outputs.BaseModelOutputWithPooling.hidden_states) here :wink:

Reply:
I spends sometime on testing on chatglm-6b. It looks a little different. It implements the transformer by itself [code here](https://huggingface.co/THUDM/chatglm-6b/blob/main/modeling_chatglm.py). 
The shape of its hidden states layer returned is [seq_len, batch, hidden_size], while normal Llama is [batch, seq_len, hidden_size]. So I guess a transpose of model_output.hidden_states[-1] should be required before torch.mean

Reply:
Thank you for your comment and for taking the time to test the chatglm-6b model. I've made the necessary adjustments to the code @guruhao 
chatglm-6b embedding quality is really high, and it has shown even better results than OpenAI's ada model. Feel free to try the code in the playground!